### PR TITLE
External dependency injection system

### PR DESCRIPTION
**Pull request by Tithugues**
### Description

What do you think about this kind of DI system please?

The current system seems to be restrictive: the definition of services should be done in PHP (which doesn't seem very readable for me), also the parameters can't be defined at the beginning (only when getting the object).

With this system, we defined the template class and it can be passed everywhere quite easily, and this may be done for other classes. Also in a second step, this would enforce the principle to use injection instead of building objects in other objects. (I would love to split the current `Response` class in `HtmlResponse` and `AjaxResponse` classes with a common `ResponseInterface`.)

Fixes #

Before submitting pull request, please review the following checklist:

- [x] Make sure you have read our [CONTRIBUTING.md](https://github.com/phpmyadmin/phpmyadmin/blob/master/CONTRIBUTING.md) document.
- [x] Make sure you are making a pull request against the correct branch. For example, for bug fixes in a released version use the corresponding QA branch and for new features use the _master_ branch. If you have a doubt, you can ask as a comment in the bug report or on the mailing list.
- [x] Every commit has proper `Signed-off-by` line as described in our [DCO](https://github.com/phpmyadmin/phpmyadmin/blob/master/DCO). This ensures that the work you're submitting is your own creation.
- [x] Every commit has a descriptive commit message.
- [x] Every commit is needed on its own, if you have just minor fixes to previous commits, you can squash them.
- [x] Any new functionality is covered by tests.

### CODE REVIEW

**Review by MauricioFauth — APPROVED** [review]

**Comment by mauriciofauth** [comment]
> I would love to split the current `Response` class in `HtmlResponse` and `AjaxResponse` classes with a common `ResponseInterface`.

Great idea. As I said in https://github.com/phpmyadmin/phpmyadmin/pull/15248#issuecomment-490535168, I also want to improve how we handle request and response with a HTTP library.

**Comment by Tithugues** [comment]
If you find any bug to fix or any code to improve, please tell me, I'll do my best to do it.

**Comment by Tithugues** [comment]
Hi,
Do you please know how to get rid of the errors thrown in https://travis-ci.org/phpmyadmin/phpmyadmin/jobs/531442853 ?

`Undefined variable: $containerBuilder`

This variable is defined in `common.inc.php` which is called at the beginning of all files which have the error. :confused: 

Thank you.

**Comment by mauriciofauth** [comment]
> This variable is defined in common.inc.php which is called at the beginning of all files which have the error. :confused:

The static analyzer doesn't know about variables defined in another files. We have been ignoring these errors.

https://github.com/phpmyadmin/phpmyadmin/blob/720e302de5c1ad5db1fb1cbcad3eaaec948ca465/phpstan.neon.dist#L18

**Comment by mauriciofauth** [comment]
Some ways to solve this would be to get the variable some other way, or if it is a global variable, declare it in the file.

**Comment by Tithugues** [comment]
> Some ways to solve this would be to get the variable some other way, or if it is a global variable, declare it in the file.

Yes, I know this is done for other variables. But doing this would encourage use of the container through the global variable in many places, whereas the idea would rather be to provide the data as an input.
So if this is OK for you, I would prefer to fill the `phpstan.neon.dist` file.

**Comment by mauriciofauth** [comment]
> So if this is OK for you, I would prefer to fill the `phpstan.neon.dist` file.

No problem.

**Comment by Tithugues** [comment]
Please don't merge it for now, I've an error after the creation of a table.

**Comment by Tithugues** [comment]
> Please don't merge it for now, I've an error after the creation of a table.

Please ignore my previous message, it was on the other PR I'm working on.

File: ajax.php [hunk 1]
```diff
@@ -33,11 +33,8 @@
 
 $response->setAjax(true);
 
-$controller = new AjaxController(
-    $response,
-    $dbi,
-    $GLOBALS['PMA_Config']
-);
+/** @var AjaxController $controller */
+$controller = $containerBuilder->get('ajax_controller');
```
Comment (mauriciofauth) [inline ajax.php]: What do you think about using `AjaxController::class` instead of `'ajax_controller'`?

Comment (Tithugues) [reply ajax.php]: This is OK for me. I'll rename the controller services with the class name of the controllers.

File: db_events.php [hunk 2]
```diff
@@ -33,11 +34,17 @@
...
-    $db
-);
+/* Define dependencies for the concerned controller */
+$dependency_definitions = [
+    'db' => $container->get('db'),
+];
+
+/** @var Definition $definition */
+$definition = $containerBuilder->getDefinition('PhpMyAdmin\Controllers\Database\EventsController');
+$definition->setArguments(array_merge($definition->getArguments(), $dependency_definitions));
```
Comment (williamdes) [inline db_events.php]: Is it possible to avoid having array merges?

File: db_multi_table_query.php [hunk 3]
```diff
@@ -29,11 +30,17 @@
...
-    $dbi,
-    $db
-);
+/* Define dependencies for the concerned controller */
+$dependency_definitions = [
+    'db' => $container->get('db'),
+];
+
+/** @var Definition $definition */
+$definition = $containerBuilder->getDefinition('PhpMyAdmin\Controllers\Database\MultiTableQueryController');
```
Comment (williamdes) [inline db_multi_table_query.php]: Can you use ::class ?

**Review by williamdes — CHANGES_REQUESTED** [review]
Good work!

File: db_events.php (hunk 2 above)
Comment (Tithugues) [reply db_events.php]: 2 things please:
1/ thank you because I wanted to replace it
2/ what would be your reason to avoid `array_merge` please?

File: db_multi_table_query.php (hunk 3 above)
Comment (Tithugues) [reply db_multi_table_query.php]: Oups… Missed this one…

File: db_events.php (hunk 2 above)
Comment (williamdes) [reply db_events.php]: It feels strange, kind of broken
I would love to see a `->setArgument('name', $value);`
And also if you invert or do a mistake in the order of an array merge the output will not be what you expected.

Comment (Tithugues) [reply db_events.php]: OK, thanks for feedback. It should be OK now.

File: db_central_columns.php [hunk 4]
```diff
@@ -33,13 +34,18 @@
...
 
-$controller = new CentralColumnsController(
-    $response,
-    $dbi,
-    $db,
-    $centralColumns
-);
+/** @var Definition $definition */
+$definition = $containerBuilder->getDefinition('database_central_columns_controller');
+$definition->setArguments(array_merge($definition->getArguments(), $dependency_definitions));
```
Comment (williamdes) [inline db_central_columns.php]: @Tithugues Can you remove `array_merge` and all the other occurrences ?

File: db_central_columns.php [hunk 5]
```diff
@@ -33,13 +34,18 @@
...
+];
 
-$controller = new CentralColumnsController(
-    $response,
-    $dbi,
-    $db,
-    $centralColumns
-);
+/** @var Definition $definition */
+$definition = $containerBuilder->getDefinition('database_central_columns_controller');
```
Comment (williamdes) [inline db_central_columns.php]: missing `::class` :)

File: db_central_columns.php [hunk 6]
```diff
@@ -33,13 +34,18 @@
...
+$definition->setArguments(array_merge($definition->getArguments(), $dependency_definitions));
+
+/** @var CentralColumnsController $controller */
+$controller = $containerBuilder->get('database_central_columns_controller');
```
Comment (williamdes) [inline db_central_columns.php]: here also

File: db_central_columns.php (hunk 4 above)
Comment (Tithugues) [reply db_central_columns.php]: Yes, I'm going to try.

File: server_databases.php [hunk 7]
```diff
@@ -12,26 +12,25 @@
...
 use PhpMyAdmin\Response;
+use Symfony\Component\DependencyInjection\ContainerInterface;
 
 if (! defined('ROOT_PATH')) {
     define('ROOT_PATH', __DIR__ . DIRECTORY_SEPARATOR);
 }
 
 require_once ROOT_PATH . 'libraries/common.inc.php';
 
+$response = $containerBuilder->get('response', ContainerInterface::NULL_ON_INVALID_REFERENCE) ?? Response::getInstance();
```
Comment (williamdes) [inline server_databases.php]: Can you use `::class` ?

**Review by williamdes — CHANGES_REQUESTED** [review]

**Review by MauricioFauth — APPROVED** [review]
Excellent work!

**Comment by mauriciofauth** [comment]
@Tithugues Do you think it's now possible to remove our custom DIC (`PhpMyAdmin\Di\Container`)?

File: server_databases.php (hunk 7 above)
Comment (Tithugues) [reply server_databases.php]: As I defined `response` as a service in the other PR, I would like to do it in the other one if this is OK for you.

File: db_central_columns.php (hunk 6 above)
Comment (Tithugues) [reply db_central_columns.php]: Done.

File: db_central_columns.php (hunk 5 above)
Comment (Tithugues) [reply db_central_columns.php]: Done.

**Comment by Tithugues** [comment]
> @Tithugues Do you think it's now possible to remove our custom DIC (`PhpMyAdmin\Di\Container`)?

Not yet. But I should confess this is one on my aims…
I think we should go further before this. Your current `Container` has one advantage that I would like not to implement with the current one: `getInstance`. Thanks to this, you can retrieve every service everywhere. But if we achieve to inject dependencies instead of creating objects, we should be able to remove the current `container`, I think.
(This is a bit confusing because I think that I'm currently doing 2 things at the same time: defining the services externally thanks to the Symfony DI + injecting dependencies.)
The first things I'm currently trying to get rid of the current `container` are:
* registering every old services as a service with the new `container`;
* replacing as much as possible the `get` of services by a `get` through the new `container`.

When all the `get` will be replaced, it won't be fully finished, because there is also some magic in your `container` that loads the parameters regarding name and/or type (don't really remember, I had a quick look) of registered services.

File: db_central_columns.php (hunk 4 above)
Comment (Tithugues) [reply db_central_columns.php]: This is done!
As there is no `replaceArguments`, I had to do it for each parameter.

Later, I think that we'll have to find a way not to duplicate this closure.

Comment (williamdes) [reply db_central_columns.php]: @Tithugues Seems like you did not push on the right branch :)

Comment (Tithugues) [reply db_central_columns.php]: Seems you're right… :sob:

File: db_datadict.php [hunk 8]
```diff
@@ -34,13 +35,19 @@
...
+/* Define dependencies for the concerned controller */
+$dependency_definitions = [
+    'db' => $container->get('db'),
+    'relation' => new Relation($dbi),
+    'transformations' => new Transformations(),
+];
+
+/** @var Definition $definition */
+$definition = $containerBuilder->getDefinition(DataDictionaryController::class);
+$definition->setArguments(array_merge($definition->getArguments(), $dependency_definitions));
```
Comment (williamdes) [inline db_datadict.php]: You missed this one :)

Comment (Tithugues) [reply db_datadict.php]: Thanks. It should be OK this time…

**Review by williamdes — APPROVED** [review]
:100:

**Comment by Tithugues** [comment]
Do you want me to squash the commits before merging the branch?

**Comment by mauriciofauth** [comment]
You can do it, if you want.

**Comment by Tithugues** [comment]
Done.

**Review by MauricioFauth — APPROVED** [review]

**Comment by mauriciofauth** [comment]
There are a few Selenium errors. You can open the link in the log to see the video of the test. 

https://travis-ci.org/phpmyadmin/phpmyadmin/jobs/534311833#L494

**Comment by Tithugues** [comment]
> There are a few Selenium errors. You can open the link in the log to see the video of the test.
> 
> https://travis-ci.org/phpmyadmin/phpmyadmin/jobs/534311833#L494

Thanks, I did some fixes. I'll wait for Selenium feedback.

**Comment by mauriciofauth** [comment]
Excellent work! Thank you for your contribution.
